sql/core/src/main/scala/org/apache/spark/sql/execution/DataSourceScanExec.scala

```diff
@@ -166,10 +166,10 @@ case class FileSourceScanExec(
     override val tableIdentifier: Option[TableIdentifier])
   extends DataSourceScanExec with ColumnarBatchScan  {
 
-  override val supportsBatch: Boolean = relation.fileFormat.supportBatch(
+  override lazy val supportsBatch: Boolean = relation.fileFormat.supportBatch(
     relation.sparkSession, StructType.fromAttributes(output))
 
-  override val needsUnsafeRowConversion: Boolean = {
+  override lazy val needsUnsafeRowConversion: Boolean = {
     if (relation.fileFormat.isInstanceOf[ParquetSource]) {
       SparkSession.getActiveSession.get.sessionState.conf.parquetVectorizedReaderEnabled
```

Comment:
Since you mentioned `SparkSession`, that line caught my attention where the active `SparkSession` is accessed using `SparkSession.getActiveSession.get` not `relation.sparkSession` as is the case for other places. I think that's something worth considering changing since we're at it.

Reply:
Let's leave this out of this PR's scope. That's more like making the plan workable whereas this PR targets the plan can be canonicalized.